app/src/main/java/emu/skyline/applet/swkbd/SoftwareKeyboardFilter.kt

```diff
@@ -0,0 +1,57 @@
+/*
+ * SPDX-License-Identifier: MPL-2.0
+ * Copyright © 2022 Skyline Team and Contributors (https://github.com/skyline-emu/)
+ */
+
+package emu.skyline.applet.swkbd
+
+import android.text.InputFilter
+import android.text.SpannableStringBuilder
+import android.text.Spanned
+
+class SoftwareKeyboardFilter(val config : SoftwareKeyboardConfig) : InputFilter {
+    override fun filter(source : CharSequence, start : Int, end : Int, dest : Spanned, dstart : Int, dend : Int) : CharSequence? {
+        val newCharacterPos = arrayOfNulls<Int>(end - start)
+        val sourceCodepoints = source.subSequence(start, end).codePoints().iterator()
+        val filteredStringBuilder = SpannableStringBuilder()
+        val replacedLength = dest.subSequence(dstart, dend).length
+        var currentChar : Int
+        var nextCharacter = start
+        while (sourceCodepoints.hasNext()) {
+            var codepoint = sourceCodepoints.next()
+            currentChar = nextCharacter
+            nextCharacter += Character.charCount(codepoint)
+            if (config.invalidCharsFlags.outsideOfDownloadCode)
+                codepoint = Character.toUpperCase(codepoint)
+            if (!config.isValid(codepoint))
+                continue
+            if (dest.length - replacedLength + filteredStringBuilder.length + Character.charCount(codepoint) > config.textMaxLength.toInt())
+                break
+            else
+                newCharacterPos[currentChar - start] = filteredStringBuilder.length
+            filteredStringBuilder.append(String(Character.toChars(codepoint)))
+        }
+        if (source is Spanned) {
+                for (span in source.getSpans(start, end, Any::class.java)) {
+                    val spanStart = source.getSpanStart(span) - start
+                    val spanFlags = source.getSpanFlags(span)
+                    val spanEnd = source.getSpanEnd(span) - start
+                    currentChar = spanStart
+                    var newStart = newCharacterPos[currentChar]
+                    while (newStart == null && currentChar < spanEnd - 1) {
+                        currentChar++
+                        newStart = newCharacterPos[currentChar]
+                    }
+                    currentChar = spanEnd - 1
+                    var newEnd = newCharacterPos[currentChar]
+                    while (newEnd == null && currentChar > spanStart) {
+                        currentChar--
+                        newEnd = newCharacterPos[currentChar]
+                    }
```

Comment:
Do the same thing above, I see no reason to not do this.
```suggestion
                    while (newEnd == null && currentChar > spanStart)
                        newEnd = newCharacterPos[--currentChar]
```